### PR TITLE
Github Actions: automatically build and upload binaries on release 

### PR DESCRIPTION
**Pull request by zyedidia**
In order to be easily usable with Chisel soon, CIRCT should provide pre-built binaries at least for macOS and Linux. Currently pre-built Linux binaries are manually uploaded whenever a new build is made. This PR adds a Github Actions script that builds and uploads macOS and Linux binaries whenever a release (or pre-release) is made (requested in #2842). The build from scratch takes approximately 2 hours.

The generated binaries are not fully static but are portable enough to run on base versions of macOS and Ubuntu/other Linuxes. The dynamic linking ends up looking like this:

Linux:

```
$ ldd firtool
	linux-vdso.so.1 (0x00007ffc10fd1000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f19d67c3000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f19d6441000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f19d6229000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f19d5e64000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f19d7142000)
```

Mac:

```
$ otool -L firtool
firtool:
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
```

You can try out the binaries from an example release [here](https://github.com/zyedidia/circt/releases/tag/test-release-15).

The build is currently set up to use gcc/g++ instead of clang because clang includes additional dynamic dependencies on Linux (`librt.so.1`, `libdl.so.2`, and `libz.so.1`) and does not support the `-static-libstdc++` flag on mac. The build currently runs on macOS-11 and Ubuntu-20.04. Are there other versions we would like to include, like Ubuntu-18.04? (I also haven't investigated running on Windows)

Currently the uploaded package contains everything in the CIRCT `bin` directory after a build. Are there other files we would like to include as well?

Also please let me know if there are other build options that would be good to enable for this. For example, I investigated making a fully static build but wasn't successful. Let me know what you think, thanks!

### CODE REVIEW

**Comment by mikeurbach** [comment]
I have a slight preference for doing a unified build of LLVM and CIRCT at the same time, although I don't know if this will actually make any difference to the final binaries. One way to do a unified build is through LLVM_EXTERNAL_PROJECTS, as documented here, for example: https://circt.llvm.org/docs/PythonBindings/#manual-compilation. A nice side effect is from scratch builds don't have to compile all of LLVM, since CMake knows exactly which libraries you depend on.

> Are there other versions we would like to include, like Ubuntu-18.04?

Ubuntu 18 wouldn't hurt, if it's easy to add to the matrix. I think there are still a good number of users on that.

> Are there other files we would like to include as well?

I was going to suggest maybe enabling the Python bindings, and including the generated Python code, but on second thought this should probably be accomplished by a different CI that just builds Python wheels, and setting that up is more involved than this.

File: .github/workflows/uploadBinaries.yml
Comment (mikeurbach) [inline .github/workflows/uploadBinaries.yml:111]: Is this required for some reason?

File: .github/workflows/uploadBinaries.yml
Comment (mikeurbach) [inline .github/workflows/uploadBinaries.yml:112]: Does LLVM_TARGETS_TO_BUILD matter for the CIRCT compile?

**Review by mikeurbach — APPROVED** [review]
Thanks for putting this together! I think the CMake options and workflow look good.

File: .github/workflows/uploadBinaries.yml
Comment (zyedidia) [reply .github/workflows/uploadBinaries.yml:111]: Yes, with the default number of parallel link jobs the macOS runner runs out of memory and crashes.

File: .github/workflows/uploadBinaries.yml
Comment (zyedidia) [reply .github/workflows/uploadBinaries.yml:112]: I'm not sure -- for that option I was just following the readme instructions, which say to use it for the LLVM build.

**Comment by zyedidia** [comment]
Great! The unified build seems nice especially if it significantly improves the compile time. I'll try that out and add ubuntu 18.04 to the matrix as well.

**Comment by zyedidia** [comment]
It seems like the unified build is similar in size to the split build, and it puts all the LLVM binaries in the `bin` directory as well (roughly 600MB zipped). Do you think we should just stick to the split build, or use the unified build and pick out the particular binaries related to CIRCT?

**Comment by zyedidia** [comment]
I'm noticing that the tar utility on the macOS runner sometimes creates corrupted archives (quite strange). For example in this run (https://github.com/zyedidia/circt/runs/7554868400?check_suite_focus=true) during the "Display Files" step it shows that everything has been successfully built:

```
build/bin/circt-capi-ir-test:     Mach-O 64-bit executable x86_64
build/bin/circt-lsp-server:       Mach-O 64-bit executable x86_64
build/bin/circt-opt:              Mach-O 64-bit executable x86_64
build/bin/circt-reduce:           Mach-O 64-bit executable x86_64
build/bin/circt-rtl-sim.py:       Python script text executable, ASCII text
build/bin/circt-translate:        Mach-O 64-bit executable x86_64
build/bin/driver.cpp:             c program text, ASCII text
build/bin/driver.sv:              ASCII text
build/bin/esi-cosim-runner.py:    Python script text executable, ASCII text
build/bin/esi-tester:             Mach-O 64-bit executable x86_64
build/bin/firtool:                Mach-O 64-bit executable x86_64
build/bin/handshake-runner:       Mach-O 64-bit executable x86_64
build/bin/llhd-sim:               Mach-O 64-bit executable x86_64
build/bin/py-split-input-file.py: Python script text executable, ASCII text
```

But then when I download the tarball I see

```
circt-capi-ir-test:     Mach-O 64-bit executable x86_64
circt-lsp-server:       data
circt-opt:              data
circt-reduce:           data
circt-rtl-sim.py:       Python script text executable, ASCII text
circt-translate:        Mach-O 64-bit executable x86_64
driver.cpp:             c program text, ASCII text
driver.sv:              ASCII text
esi-cosim-runner.py:    Python script text executable, ASCII text
esi-tester:             Mach-O 64-bit executable x86_64
firtool:                data
handshake-runner:       Mach-O 64-bit executable x86_64
llhd-sim:               data
py-split-input-file.py: Python script text executable, ASCII text
```

The tarball I downloaded has the same hash as the one created on the server before upload so it appears to be corruption happening during the tar creation rather than during upload. For some builds it works fine though.

I'll look into why that's happening.

**Comment by zyedidia** [comment]
Apparently there is a bug in tar on macOS runners that is particularly noticeable when bundling LLVM binaries. See https://github.com/actions/virtual-environments/issues/2619. I think the fix is to use GNU tar on mac so I have switched the runner to use that locally and am running a test.

**Comment by zyedidia** [comment]
I added Ubuntu 18.04 to the matrix, and using GNU tar on macOS fixed the tar problem, so I think that this is ready to go unless you strongly prefer doing a unified build (but in that case I'm not sure if a subset of the binaries should be selected).

**Comment by mikeurbach** [comment]
> unless you strongly prefer doing a unified build

I don't strongly prefer it, and having LLVM binaries compiled unnecessarily is good enough reason to not worry about that. Thanks for taking a look.

**Review by mikeurbach — APPROVED** [review]

**Comment by seldridge** [comment]
> Currently the uploaded package contains everything in the CIRCT bin directory after a build. Are there other files we would like to include as well?

I'm not sure, but a safe bet would be anything created as part of the install build target, i.e., `bin/`, `include/`, `lib/`, and `lib64/`.

Also, this is awesome! Thanks for tackling this.

Future improvements would be to provide a rpm/deb as opposed to an archive. Not a huge deal though as this is immediately usable as-is.

**Review by youngar — APPROVED** [review]
LGTM.  @zyedidia if you don't have merge permissions let us know when it's ready to merge and someone will hit the button for you 👍 . I plan on doing a release today so if we can merge this in the morning we can see if it works as tested.

**Comment by zyedidia** [comment]
Awesome! I think this is ready to merge and I don't have merge permissions. For this release we should probably also include the normal `circt-release.tgz` and then afterwards the one chipfactory image that uses it can be switched to the new release tarballs and `circt-release.tgz` won't have to be uploaded anymore.

**Comment by youngar** [comment]
Thanks @zyedidia!

**Pull request merged**
